File: baekjoon/Python/9663_N-Queen.py
N = int(input())
ans = 0


def dfs(row):
    global ans
    if row == N:
        ans += 1
        return

    for j in range(N):  # 특정행의 열 방문
        # 같은 열, 대각선에 방문 X
        if not v1[j] and not v2[row + j] and not v3[row - j + N]:  # 조건을 and로 연결
            v1[j] = v2[row + j] = v3[row - j + N] = True  # 값을 True로 설정
            dfs(row + 1)
            v1[j] = v2[row + j] = v3[row - j + N] = False  # 값을 False로 재설정


v1 = [False] * N
v2 = [False] * (2 * N)
v3 = [False] * (2 * N)
dfs(0)
print(ans)

File: baekjoon/Java/9663_N-Queen.java
import java.util.Scanner;

/**
* N - Queen
* */
public class P_9663 {
    static boolean[] v1;
    static boolean[] v2;
    static boolean[] v3;
    static int ans;
    static int N;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();

        v1=new boolean[N];
        v2=new boolean[N*2];
        v3=new boolean[N*2];

        ans = 0;
        dfs(0);
        System.out.println(ans);
    }
    public static void dfs(int row){
        // 종료조건
        if(row == N){
            ans+=1;
            return;
        }
        for(int j = 0; j < N; j++){
            if(v1[j]==false && v2[j+row]==false && v3[row-j+N]==false){
                v1[j]=true;
                v2[j+row]=true;
                v3[row-j+N]=true;
                dfs(row + 1);
                v1[j]=false;
                v2[j+row]=false;
                v3[row-j+N]=false;
            }
        }
    }
}
